File: LeetCode/Python/Medium/1570-Dot-Product-of-Two-Sparse-Vectors/solution.py
class SparseVector:
    def __init__(self, nums: List[int]):
        self.vector=nums

    def get_vector(self) -> List[int]:
        return self.vector

    # Return the dotProduct of two sparse vectors
    def dotProduct(self, vec: 'SparseVector') -> int:
        dot_product=0
        
        vector1=self.get_vector()

        vector2=vec.get_vector()
        
        n=len(vector1)

        for index in range(0, n):
            product=vector1[index] * vector2[index]

            dot_product+=product

        return dot_product

class SparseVector:
    def __init__(self, nums: List[int]):
        self.non_zero_hashmap=dict()

        for (index, number) in enumerate(nums):
            if number > 0:
                self.non_zero_hashmap[index]=number

    def get_non_zero_hashmap(self) -> Dict[int, int]:
        return self.non_zero_hashmap

    # Return the dotProduct of two sparse vectors
    def dotProduct(self, vec: 'SparseVector') -> int:
        dot_product=0
        
        vector1_non_zero_hashmap=self.get_non_zero_hashmap()

        vector2_non_zero_hashmap=vec.get_non_zero_hashmap()

        for (index, number) in vector1_non_zero_hashmap.items():
            if index in vector2_non_zero_hashmap.keys():
                product=vector1_non_zero_hashmap[index] * vector2_non_zero_hashmap[index]

                dot_product+=product

        return dot_product

File: LeetCode/Rust/Medium/1570-Dot-Product-of-Two-Sparse-Vectors/solution.rs
use std::collections::{HashMap};

struct SparseVector
{
	non_zero_hashmap: HashMap<usize, i32>
}

/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl SparseVector
{
    fn new(nums: Vec<i32>) -> Self
    {   
        let mut _non_zero_hashmap: HashMap<usize, i32>=nums.into_iter().enumerate().filter(|(index, number)| *number!=0).collect::<HashMap<usize, i32>>();

        Self
        {
            non_zero_hashmap: _non_zero_hashmap
        }
    }
	
    // Return the dotProduct of two sparse vectors
    fn dot_product(&self, vec: SparseVector) -> i32
    {
        let mut dot_product: i32=0;

        for (index, number) in self.non_zero_hashmap.iter()
        {
            match (vec.non_zero_hashmap.get(index))
            {
                Some(value) =>
                {
                    let product: i32=(*number) * value;

                    dot_product+=product;
                },
                None => ()
            }
        }

        dot_product
    }
}